rclpy/src/rclpy/wait_set.cpp

```diff
@@ -0,0 +1,296 @@
+    if (!event) {
+      throw py::error_already_set();
+    }
+    ret = rcl_wait_set_add_event(wait_set, event, &index);
+  } else {
+    std::string error_text{"'"};
+    error_text += entity_type;
+    error_text += "' is not a known entity";
+    throw std::runtime_error(error_text);
+  }
+  if (ret != RCL_RET_OK) {
+    std::string error_text{"failed to add'"};
+    error_text += entity_type;
+    error_text += "' to waitset";
+    throw RCLError(error_text);
+  }
+  return index;
+}
+
+bool
+wait_set_is_ready(const std::string entity_type, py::capsule pywait_set, size_t index)
+{
+  if (0 != std::strcmp("rcl_wait_set_t", pywait_set.name())) {
+    throw py::value_error("capsule is not an rcl_wait_set_t");
+  }
+  auto wait_set = static_cast<rcl_wait_set_t *>(pywait_set);
+
+  const void ** entities = NULL;
+  size_t num_entities = 0;
+  if ("subscription" == entity_type) {
+    entities = reinterpret_cast<const void **>(wait_set->subscriptions);
+    num_entities = wait_set->size_of_subscriptions;
+  } else if ("client" == entity_type) {
+    entities = reinterpret_cast<const void **>(wait_set->clients);
+    num_entities = wait_set->size_of_clients;
+  } else if ("service" == entity_type) {
+    entities = reinterpret_cast<const void **>(wait_set->services);
+    num_entities = wait_set->size_of_services;
+  } else if ("timer" == entity_type) {
+    entities = reinterpret_cast<const void **>(wait_set->timers);
+    num_entities = wait_set->size_of_timers;
+  } else if ("guard_condition" == entity_type) {
+    entities = reinterpret_cast<const void **>(wait_set->guard_conditions);
+    num_entities = wait_set->size_of_guard_conditions;
+  } else if ("event" == entity_type) {
+    entities = reinterpret_cast<const void **>(wait_set->events);
+    num_entities = wait_set->size_of_events;
+  } else {
+    std::string error_text{"'"};
+    error_text += entity_type;
+    error_text += "' is not a known entity";
+    throw std::runtime_error(error_text);
+  }
+
+  if (!entities) {
+    std::string error_text{"wait set '"};
+    error_text += entity_type;
+    error_text += "' isn't allocated";
+    throw std::runtime_error(error_text);
+  }
+  if (index >= num_entities) {
+    throw std::out_of_range("wait set index too big");
+  }
+  return nullptr != entities[index];
+}
+
+template<typename EntityArray>
+py::list
+_get_ready_entities(const EntityArray ** entities, const size_t num_entities)
+{
+  py::list entity_list;
+  for (size_t i = 0; i < num_entities; ++i) {
+    auto address = reinterpret_cast<size_t>(entities[i]);
+    if (address) {
+      entity_list.append(address);
+    }
+  }
+  return entity_list;
+}
+
+py::list
+get_ready_entities(const std::string entity_type, py::capsule pywait_set)
+{
+  if (0 != std::strcmp("rcl_wait_set_t", pywait_set.name())) {
+    throw py::value_error("capsule is not an rcl_wait_set_t");
+  }
+  auto wait_set = static_cast<rcl_wait_set_t *>(pywait_set);
+
+  py::list entity_ready_list;
```

Comment:
I think this can be dropped.
```suggestion
```

Reply:
Removed unused variable in 3121085 (didn't use gh suggestions to make DCO bot happy)